src/wallet/scriptpubkeyman.h

```diff
@@ -0,0 +1,484 @@
+
+    //! Upgrade stored CKeyMetadata objects to store key origin info as KeyOriginInfo
+    virtual void UpgradeKeyMetadata() {}
+
+    /** Sets up the key generation stuff, i.e. generates new HD seeds and sets them as active.
+      * Returns false if already setup or setup fails, true if setup is successful
+      * Set force=true to make it re-setup if already setup, used for upgrades
+      */
+    virtual bool SetupGeneration(bool force = false) { return false; }
+
+    /* Returns true if HD is enabled */
+    virtual bool IsHDEnabled() const { return false; }
+
+    /* Returns true if the wallet can give out new addresses. This means it has keys in the keypool or can generate new keys */
+    virtual bool CanGetAddresses(bool internal = false) { return false; }
+
+    /** Upgrades the wallet to the specified version */
+    virtual bool Upgrade(int prev_version, int new_version, std::string& error) { return false; }
+
+    virtual bool HavePrivateKeys() const { return false; }
+
+    //! The action to do when the DB needs rewrite
+    virtual void RewriteDB() {}
+
+    virtual int64_t GetOldestKeyPoolTime() { return GetTime(); }
+
+    virtual size_t KeypoolCountExternalKeys() { return 0; }
+    virtual unsigned int GetKeypoolSize() const { return 0; }
+
+    virtual int64_t GetTimeFirstKey() const { return 0; }
+
+    virtual const CKeyMetadata* GetMetadata(uint160 id) const { return nullptr; }
+
+    virtual std::unique_ptr<SigningProvider> GetSigningProvider(const CScript& script) const { return nullptr; }
+
+    /** Whether this ScriptPubKeyMan can provide a SigningProvider (via GetSigningProvider) that, combined with
+      * sigdata, can produce a valid signature.
+      */
+    virtual bool CanProvide(const CScript& script, SignatureData& sigdata) { return false; }
+
+    virtual uint256 GetID() const { return uint256(); }
+
+    /** Prepends the wallet name in logging output to ease debugging in multi-wallet use cases */
+    template<typename... Params>
+    void WalletLogPrintf(std::string fmt, Params... parameters) const {
+        LogPrintf(("%s " + fmt).c_str(), GetDisplayName(), parameters...);
+    };
+
+    /** Watch-only address added */
+    boost::signals2::signal<void (bool fHaveWatchOnly)> NotifyWatchonlyChanged;
+
+    /** Keypool has new keys */
+    boost::signals2::signal<void ()> NotifyCanGetAddressesChanged;
+};
+
+class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProvider
+{
+private:
+    //! if fUseCrypto is true, mapKeys must be empty
+    //! if fUseCrypto is false, vMasterKey must be empty
+    std::atomic<bool> fUseCrypto{false};
+
+    bool SetCrypted();
+
+    //! keeps track of whether Unlock has run a thorough check before
+    bool fDecryptionThoroughlyChecked = false;
+
+    //! Master encryption key for the wallet. Keys are encrypted with this.
+    //! Wallet is unlocked if this is not null.
+    CKeyingMaterial vMasterKey GUARDED_BY(cs_KeyStore);
```

Comment:
I've added `WalletStorage` as you suggested. However I did not change `vMasterKey`, `fUseCrypto`, and `fDecryptionThorouglyChecked` to be references or pointers as allowing `ScriptPubKeyMan`s to modify these variables in `CWallet` itself may result in inconsistent states where one `ScriptPubKeyMan` has finished en/decrypting but another has not. Properly changing them to use references could be done in the future, but doing it now would cause the places that use them to differ even more and thus make review harder.

Reply:
> However, at the PR level, shouldn't `CWallet`'s `IsCrypted()` be looping over all `ScriptPubKeyMan`s, just like `SetCrypted()` does?

`SetCrypted` doesn't loop.

The thought I had here was that it could be possible that a wallet is considered encrypted (by virtue of having an encryption key set) without any of the `ScriptPubKeyMan`s having any of them actually encrypting anything and thus be `IsCrypted() == false`.

Reply:
The encryption stuff here is making my head spin. Wallet encryption was already confusing before, but now with CWallet and ScriptPubKeyMan both having duplicate `fUseCrypto`, `SetCrypted()`, `IsCrypted()`,  `vMasterKey`, `IsLocked()` states and code, it's hard to tell which combinations of state are expected, and which are bugs, and whether the code supposed to be keeping states in sync actually works.

I have a few suggestions to try clarify things. These should also simplify the code.

1. Drop the `CWallet::fUseCrypto` member and `CWallet::SetCrypted()` method. Change `CWallet::IsCrypted()` method to just loop over keyman objects and call IsCrypted on them. As far as I can tell the `CWallet::fUseCrypto` state is broken in the current PR. `CWallet::SetCrypted()` used to be called whenever any encrypted data was present, but now that now longer happens, and `CWallet::IsCrypted()` returns false if the wallet hasn't been explicitly encrypted or locked since startup, regardless of what encrypted keys or data are present.

2. If the intent is to consider a wallet encrypted "by virtue of having an encryption key set", it'd be better to add a new method `bool HasEncryptionKey() { return !mapCryptedKeys.empty(); }` instead of overloading `IsCrypted()`. It might also make sense to rename the existing `IsCrypted()` function `HasEncryptedData()` in that case so the meaning is not confused.

3. This last suggestion is less important, but I think it'd be better to lock and unlock the entire wallet, instead of parts of it. The keyman `vMasterKey` member and `IsLocked()`, `Locked()`, and `Unlocked()` implementations copied from wallet code could all be dropped. Keyman instances should have a `fDecryptionThoroughlyChecked` member and have `CheckDecryption()` methods, but otherwise would be simpler if they just called `IsLocked()` and `GetMasterKey()` from `WalletStorage` instead of implementing locking and unlocking themselves.

Reply:
I'll take a look at reworking the encryption state stuff.

Reply:
I've changed the encryption stuff to be mostly in line with @ryanofsky's suggestions.

***

> As far as I can tell the `CWallet::fUseCrypto` state is broken in the current PR. `CWallet::SetCrypted()` used to be called whenever any encrypted data was present, but now that now longer happens, and `CWallet::IsCrypted()` returns false if the wallet hasn't been explicitly encrypted or locked since startup, regardless of what encrypted keys or data are present.

FWIW it wasn't broken. `SetCrypted` was changed to be called during wallet loading so `fUseCrypto` would be true and thus `IsCrypted()` returned true. Otherwise tests wouldn't pass.